src/kits/tracker/FindPanel.cpp

```diff
@@ -262,7 +282,222 @@ FindWindow::FindWindow(const entry_ref* newRef, bool editIfTemplateOnly)
+		new BMessage(kClearTemplates)));
+	fOptionsMenu->AddItem(new BMenuItem(B_TRANSLATE("Clear History"),
+		new BMessage(kClearHistory)));
+
+	PopulateTemplatesMenu();
+
+	fMenuBar->AddItem(fQueryMenu);
+	fMenuBar->AddItem(fOptionsMenu);
+	fMenuBar->AddItem(fTemplatesMenu);
+}
+
+
+void
+FindWindow::UpdateFileReferences(const entry_ref* ref)
+{
+	BEntry entry(ref);
+	if(!entry.Exists())
+		return;
+
+	fFile->Unset();
+	fFile = NULL;
+	fFile = TryOpening(ref);
+
+	if(fFile != NULL) {
+		entry.GetRef(&fRef);
+		fFromTemplate = IsQueryTemplate(fFile);
+	}
+}
+
+
+void
+FindWindow::ClearMenu(BMenu* menu)
+{
+
+	int32 count = menu->CountItems();
+	for(int32 i = 0;i<count;i++) {
+		BMenuItem* item = menu->RemoveItem(0);
+		delete item;
+	}
+}
+
+void
+FindWindow::DeleteQueryOrTemplate(BEntry* entry)
+{
+	if(entry->Exists()) {
+		entry_ref ref;
+		entry->GetRef(&ref);
+		if(fRef == ref) {
+			UpdateFileReferences(NULL);
+			fSaveQueryOrTemplateItem->SetEnabled(false);
+		}
+		
+		entry->Remove();
+	}
+}
+
+void
+FindWindow::ClearHistoryOrTemplates(bool clearTemplates, bool temporaryOnly)
+{
+	BVolumeRoster roster;
+	BVolume volume;
+	while(roster.GetNextVolume(&volume) == B_OK) {
+		if(volume.IsPersistent() && volume.KnowsQuery() && volume.KnowsAttr()) {
+			BQuery query;
+			query.SetVolume(&volume);
+			query.SetPredicate("_trk/recentQuery == 1");
+			if(query.Fetch() != B_OK)
+				continue;
+
+			BEntry entry;
+			entry_ref ref;
+			while(query.GetNextEntry(&entry) == B_OK) {
+				entry.GetRef(&ref);
+				if(FSInTrashDir(&ref) && !BEntry(&ref).Exists())
+					continue;
+				char type[B_MIME_TYPE_LENGTH];
+				BNodeInfo(new BNode(&entry)).GetType(type);
+				if(strcmp(type, B_QUERY_TEMPLATE_MIMETYPE) == 0) {
+					if(clearTemplates)
+						DeleteQueryOrTemplate(&entry);
+					else
+						continue;
+				}
+
+				if(!clearTemplates) {
+					BFile file(&entry, B_READ_ONLY);
+					bool isTemporary;
+					if(file.ReadAttr("_trk/temporary", B_BOOL_TYPE, 0, &isTemporary,
+						sizeof(isTemporary)) == sizeof(isTemporary)) {
+						if(!temporaryOnly)
+							DeleteQueryOrTemplate(&entry);
+						else{
+							if(isTemporary)
+								DeleteQueryOrTemplate(&entry);
+						}
+					}
+				}
+			}
+		}
+	}
+
+	if(clearTemplates) {
+		ClearMenu(fTemplatesMenu);
+	} else {
+		ClearMenu(fUnsavedHistoryMenu);
+		ClearMenu(fSavedHistoryMenu);
+		FindPanel::AddRecentQueries(fUnsavedHistoryMenu, false, new BMessenger(fBackground),
+			kSwitchToQueryTemplate, false, true, false);
+		FindPanel::AddRecentQueries(fSavedHistoryMenu, false, new BMessenger(fBackground),
+			kSwitchToQueryTemplate, false, false, true);
+	}
+}
+
+
+bool
+CheckForDuplicates(BObjectList<entry_ref>* list, entry_ref* ref)
+{
+	// Simple Helper Function To Check For Duplicates Within an Entry List of Templates
+	int32 count = list->CountItems();
+	for(int32 i = 0;i<count;i++) {
+		if(strcmp(BPath(list->ItemAt(i)).Path(), BPath(ref).Path()) == 0) {
```

Comment:
BPath has implemented the == operator (see https://www.haiku-os.org/docs/api/classBPath.html#a3df539180ce332bebd6aa5b657b76254)

Reply:
Yup. I fixed this in the latest commit.